pkg/trie/triedb/iterator.go

```diff
@@ -3,14 +3,44 @@
 
 package triedb
 
+import "bytes"
+
 // Entries returns all the key-value pairs in the trie as a map of keys to values
 // where the keys are encoded in Little Endian.
 func (t *TrieDB) Entries() (keyValueMap map[string][]byte) {
-	panic("not implemented yet")
+	entries := make(map[string][]byte)
+
+	iter := NewTrieDBIterator(t)
+	for entry := iter.NextEntry(); entry != nil; entry = iter.NextEntry() {
+		entries[string(entry.key)] = entry.value
+	}
+
+	return entries
 }
 
 // NextKey returns the next key in the trie in lexicographic order.
 // It returns nil if no next key is found.
 func (t *TrieDB) NextKey(key []byte) []byte {
-	panic("not implemented yet")
+	iter := NewTrieDBIterator(t)
+
+	iter.Seek(key)
+	nextKey := iter.NextKey()
+	return nextKey
+}
+
+// GetKeysWithPrefix returns all keys in little Endian
+// format from nodes in the trie that have the given little
+// Endian formatted prefix in their key.
+func (t *TrieDB) GetKeysWithPrefix(prefix []byte) (keysLE [][]byte) {
+	iter := NewTrieDBIterator(t)
+	iter.Seek(prefix)
+
+	//Since seek consumes the prefix, we need to add it in the keys list
+	keys := [][]byte{prefix}
+
+	for key := iter.NextKey(); bytes.HasPrefix(key, prefix); key = iter.NextKey() {
+		keys = append(keys, key)
```

Comment:
is it possible that in the set of returned keys there is a key that has the prefix but is **after** the key w/o the prefix? I'm asking that because the for loop ends when the key does not have the prefix

Reply:
I don't think so, for example, if we have this trie:

![image](https://github.com/ChainSafe/gossamer/assets/659400/abc0f51d-0805-4852-818c-3fe673223b48)

And we want all the keys that share the prefix: `not`

We are first moving the iterator to the node with `key = not` using `iter.Seek(prefix)`:

![image](https://github.com/ChainSafe/gossamer/assets/659400/a5aa0b76-fc6a-4ef8-9de8-b31e9444d7fc)

- gray: nodes we already visited
- green: node we are currently visiting
- blue: pendings nodes to visit

We are going to visit "notable" and "notification" but we are going to stop visiting when the prefix is not shared anymore, for example for "test" node

Reply:
One think I noticed now is that it could be expensive to skip a large amount of keys to seek the one we are looking for, we can change that with a more smart `Seek` method, I'm going to try to address it today

Reply:
Improved! now we are traversing the trie going directly to the node with the prefix and creating an iterator from that node